README.md
# 詳説ディープラーニング（生成モデル編）

ディープラーニング電子書籍「詳説ディープラーニング（生成モデル編）」の中で紹介しているコード集です。書籍は[noteのページ](https://note.mu/yusugomori/n/n945f51cabc03)もしくは[amazon](https://amzn.to/2Dryztq)から購入ができます。

※書籍の修正履歴は[こちら](https://github.com/yusugomori/dl-book-generative/blob/master/CHANGELOG.md)にまとめています。


CHANGELOG.md
# 書籍バージョン履歴

**ver. 1.3.1 (2019.04.28)**

* p20 `mnist_test` を読み込むべきところが `mnist_train` となっていた部分を修正しました。それに伴い、実験結果の画像を差し替えています。
* p58 「識別器 G」と記述されていた部分を修正しました。

<br>


**ver. 1.3.0 (2019.04.24)**

* p9 式(1.9) KLダイバージェンスの添字の誤りを修正しました。

<br>

**ver. 1.2.0 (2019.04.19)**

* p30 (L5) 生成モデルの数式の誤植を修正しました。

<br>

**ver. 1.1.0 (2019.04.18)**

* p50 「負の交差エントロピー誤差」の部分を削除しました。
* p52 式(3.14) の添字の誤りを修正しました。

<br>

**ver. 1.0.1 (2019.04.16)**

* ３章のセクションレベルを調整しました。本文の修正はありません。

<br>

**ver. 1.0.0 (2019.04.16)**


src/02_vae_fashion_mnist.py
import os
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optimizers
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import torchvision
import torchvision.transforms as transforms
import matplotlib
# matplotlib.use('Agg')
import matplotlib.pyplot as plt


class VAE(nn.Module):
    def __init__(self, device='cpu'):
        super().__init__()
        self.device = device
        self.encoder = Encoder(device=device)
        self.decoder = Decoder(device=device)

    def forward(self, x):
        mean, var = self.encoder(x)
        z = self.reparameterize(mean, var)
        y = self.decoder(z)

        return y

    def reparameterize(self, mean, var):
        eps = torch.randn(mean.size()).to(self.device)
        z = mean + torch.sqrt(var) * eps
        return z

    def lower_bound(self, x):
        mean, var = self.encoder(x)
        kl = - 1/2 * torch.mean(torch.sum(1
                                          + torch.log(var)
                                          - mean**2
                                          - var, dim=1))
        z = self.reparameterize(mean, var)
        y = self.decoder(z)

        reconst = torch.mean(torch.sum(x * torch.log(y)
                                       + (1 - x) * torch.log(1 - y),
                                       dim=1))

        L = reconst - kl

        return L


class Encoder(nn.Module):
    def __init__(self, device='cpu'):
        super().__init__()
        self.device = device
        self.l1 = nn.Linear(784, 200)
        self.l2 = nn.Linear(200, 200)
        self.l_mean = nn.Linear(200, 10)
        self.l_var = nn.Linear(200, 10)

    def forward(self, x):
        h = self.l1(x)
        h = torch.relu(h)
        h = self.l2(h)
        h = torch.relu(h)

        mean = self.l_mean(h)
        var = F.softplus(self.l_var(h))

        return mean, var


class Decoder(nn.Module):
    def __init__(self, device='cpu'):
        super().__init__()
        self.device = device
        self.l1 = nn.Linear(10, 200)
        self.l2 = nn.Linear(200, 200)
        self.out = nn.Linear(200, 784)

    def forward(self, x):
        h = self.l1(x)
        h = torch.relu(h)
        h = self.l2(h)
        h = torch.relu(h)
        h = self.out(h)
        y = torch.sigmoid(h)

        return y


if __name__ == '__main__':
    np.random.seed(1234)
    torch.manual_seed(1234)
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    '''
    1. Load data
    '''
    root = os.path.join(os.path.dirname(__file__),
                        '.', 'data', 'fashion_mnist')
    transform = transforms.Compose([transforms.ToTensor(),
                                    lambda x: x.view(-1)])
    mnist_train = \
        torchvision.datasets.FashionMNIST(root=root,
                                          download=True,
                                          train=True,
                                          transform=transform)
    train_dataloader = DataLoader(mnist_train,
                                  batch_size=100,
                                  shuffle=True)

    '''
    2. Build model
    '''
    model = VAE(device=device).to(device)

    '''
    3. Train model
    '''
    criterion = model.lower_bound
    optimizer = optimizers.Adam(model.parameters())

    def compute_loss(x):
        return -1 * criterion(x)

    def train_step(x):
        model.train()
        loss = compute_loss(x)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        return loss

    epochs = 10

    for epoch in range(epochs):
        train_loss = 0.

        for (x, _) in train_dataloader:
            x = x.to(device)
            loss = train_step(x)

            train_loss += loss.item()

        train_loss /= len(train_dataloader)

        print('Epoch: {}, Cost: {:.3f}'.format(
            epoch+1,
            train_loss
        ))

    '''
    4. Test model
    '''
    def gen_noise(batch_size):
        return torch.empty(batch_size, 10).normal_().to(device)

    def generate(batch_size=16):
        model.eval()
        z = gen_noise(batch_size)
        gen = model.decoder(z)
        gen = gen.view(-1, 28, 28)

        return gen

    images = generate(batch_size=16)
    images = images.squeeze().detach().cpu().numpy()
    plt.figure(figsize=(6, 6))
    for i, image in enumerate(images):
        plt.subplot(4, 4, i+1)
        plt.imshow(image, cmap='binary_r')
        plt.axis('off')
    plt.tight_layout()
    plt.show()


src/01_autoencoder_fashion_mnist.py
import os
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optimizers
from torch.utils.data import Dataset, DataLoader
import torchvision
import torchvision.transforms as transforms
import matplotlib
# matplotlib.use('Agg')
import matplotlib.pyplot as plt


class Autoencoder(nn.Module):
    def __init__(self, device='cpu'):
        super().__init__()
        self.device = device
        self.l1 = nn.Linear(784, 200)
        self.l2 = nn.Linear(200, 784)

    def forward(self, x):
        # encode
        h = self.l1(x)
        h = torch.relu(h)

        # decode
        h = self.l2(h)
        y = torch.sigmoid(h)

        return y


if __name__ == '__main__':
    np.random.seed(1234)
    torch.manual_seed(1234)
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    '''
    1. Load data
    '''
    root = os.path.join(os.path.dirname(__file__),
                        '.', 'data', 'fashion_mnist')
    transform = transforms.Compose([transforms.ToTensor(),
                                    lambda x: x.view(-1)])
    mnist_train = \
        torchvision.datasets.FashionMNIST(root=root,
                                          download=True,
                                          train=True,
                                          transform=transform)
    mnist_test = \
        torchvision.datasets.FashionMNIST(root=root,
                                          download=True,
                                          train=False,
                                          transform=transform)
    train_dataloader = DataLoader(mnist_train,
                                  batch_size=100,
                                  shuffle=True)
    test_dataloader = DataLoader(mnist_test,
                                 batch_size=1,
                                 shuffle=False)

    '''
    2. Build model
    '''
    model = Autoencoder(device=device).to(device)

    '''
    3. Train model
    '''
    criterion = nn.BCELoss()
    optimizer = optimizers.Adam(model.parameters())

    def compute_loss(x, preds):
        return criterion(preds, x)

    def train_step(x):
        model.train()
        preds = model(x)
        loss = compute_loss(x, preds)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        return loss

    epochs = 10

    for epoch in range(epochs):
        train_loss = 0.

        for (x, _) in train_dataloader:
            x = x.to(device)
            loss = train_step(x)

            train_loss += loss.item()

        train_loss /= len(train_dataloader)

        print('Epoch: {}, Cost: {:.3f}'.format(
            epoch+1,
            train_loss
        ))

    '''
    4. Test model
    '''
    x, _ = next(iter(test_dataloader))
    noise = torch.bernoulli(0.8 * torch.ones(x.size())).to(device)
    x_noise = x * noise

    x_reconstructed = model(x_noise)

    plt.figure(figsize=(18, 6))
    for i, image in enumerate([x, x_noise, x_reconstructed]):
        image = image.view(28, 28).detach().cpu().numpy()
        plt.subplot(1, 3, i+1)
        plt.imshow(image, cmap='binary_r')
        plt.axis('off')
    plt.show()


src/03_gan_fashion_mnist.py
import os
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optimizers
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import torchvision
import torchvision.transforms as transforms
import matplotlib
# matplotlib.use('Agg')
import matplotlib.pyplot as plt


class GAN(nn.Module):
    def __init__(self, device='cpu'):
        super().__init__()
        self.device = device
        self.G = Generator(device=device)
        self.D = Discriminator(device=device)

    def forward(self, x):
        x = self.G(x)
        y = self.D(x)

        return y


class Discriminator(nn.Module):
    def __init__(self, device='cpu'):
        super().__init__()
        self.device = device
        self.conv1 = nn.Conv2d(1, 128,
                               kernel_size=(3, 3),
                               stride=(2, 2),
                               padding=1)
        self.relu1 = nn.LeakyReLU(0.2)
        self.conv2 = nn.Conv2d(128, 256,
                               kernel_size=(3, 3),
                               stride=(2, 2),
                               padding=1)
        self.bn2 = nn.BatchNorm2d(256)
        self.relu2 = nn.LeakyReLU(0.2)
        self.fc = nn.Linear(256*7*7, 1024)
        self.bn3 = nn.BatchNorm1d(1024)
        self.relu3 = nn.LeakyReLU(0.2)
        self.out = nn.Linear(1024, 1)

    def forward(self, x):
        h = self.conv1(x)
        h = self.relu1(h)
        h = self.conv2(h)
        h = self.bn2(h)
        h = self.relu2(h)
        h = h.view(-1, 256*7*7)
        h = self.fc(h)
        h = self.bn3(h)
        h = self.relu3(h)
        h = self.out(h)
        y = torch.sigmoid(h)

        return y


class Generator(nn.Module):
    def __init__(self,
                 input_dim=100,
                 device='cpu'):
        super().__init__()
        self.device = device
        self.linear = nn.Linear(input_dim, 256*14*14)
        self.bn1 = nn.BatchNorm1d(256*14*14)
        self.relu1 = nn.ReLU()
        self.conv1 = nn.Conv2d(256, 128,
                               kernel_size=(3, 3),
                               padding=1)
        self.bn2 = nn.BatchNorm2d(128)
        self.relu2 = nn.ReLU()
        self.conv2 = nn.Conv2d(128, 64,
                               kernel_size=(3, 3),
                               padding=1)
        self.bn3 = nn.BatchNorm2d(64)
        self.relu3 = nn.ReLU()
        self.conv3 = nn.Conv2d(64, 1,
                               kernel_size=(1, 1))

    def forward(self, x):
        h = self.linear(x)
        h = self.bn1(h)
        h = self.relu1(h)
        h = h.view(-1, 256, 14, 14)
        h = F.interpolate(h, size=(28, 28))
        h = self.conv1(h)
        h = self.bn2(h)
        h = self.relu2(h)
        h = self.conv2(h)
        h = self.bn3(h)
        h = self.relu3(h)
        h = self.conv3(h)
        y = torch.sigmoid(h)

        return y


if __name__ == '__main__':
    np.random.seed(1234)
    torch.manual_seed(1234)
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    '''
    1. Load data
    '''
    root = os.path.join(os.path.dirname(__file__),
                        '.', 'data', 'fashion_mnist')
    transform = transforms.Compose([transforms.ToTensor()])
    mnist_train = \
        torchvision.datasets.FashionMNIST(root=root,
                                          download=True,
                                          train=True,
                                          transform=transform)
    train_dataloader = DataLoader(mnist_train,
                                  batch_size=100,
                                  shuffle=True)

    '''
    2. Build model
    '''
    model = GAN(device=device).to(device)

    def gen_noise(batch_size):
        return torch.empty(batch_size, 100).uniform_(0, 1).to(device)

    '''
    3. Train model
    '''
    criterion = nn.BCELoss()
    optimizer_D = optimizers.Adam(model.D.parameters(), lr=0.0002)
    optimizer_G = optimizers.Adam(model.G.parameters(), lr=0.0002)

    def compute_loss(label, preds):
        return criterion(preds, label)

    def train_step(x):
        batch_size = x.size(0)
        model.D.train()
        model.G.train()

        # 3-A. train D
        # real images
        preds = model.D(x).squeeze()  # preds with true images
        t = torch.ones(batch_size).float().to(device)
        loss_D_real = compute_loss(t, preds)
        # fake images
        noise = gen_noise(batch_size)
        gen = model.G(noise)
        preds = model.D(gen.detach()).squeeze()  # preds with fake images
        t = torch.zeros(batch_size).float().to(device)
        loss_D_fake = compute_loss(t, preds)

        loss_D = loss_D_real + loss_D_fake
        optimizer_D.zero_grad()
        loss_D.backward()
        optimizer_D.step()

        # 3-B. train G
        noise = gen_noise(batch_size)
        gen = model.G(noise)
        preds = model.D(gen).squeeze()  # preds with fake images
        t = torch.ones(batch_size).float().to(device)  # label as true
        loss_G = compute_loss(t, preds)
        optimizer_G.zero_grad()
        loss_G.backward()
        optimizer_G.step()

        return loss_D, loss_G

    epochs = 20

    for epoch in range(epochs):
        train_loss_D = 0.
        train_loss_G = 0.
        test_loss = 0.

        for (x, _) in train_dataloader:
            x = x.to(device)
            loss_D, loss_G = train_step(x)

            train_loss_D += loss_D.item()
            train_loss_G += loss_G.item()

        train_loss_D /= len(train_dataloader)
        train_loss_G /= len(train_dataloader)

        print('Epoch: {}, D Cost: {:.3f}, G Cost: {:.3f}'.format(
            epoch+1,
            train_loss_D,
            train_loss_G
        ))

    '''
    4. Test model
    '''
    def generate(batch_size=16):
        model.eval()
        noise = gen_noise(batch_size)
        gen = model.G(noise)

        return gen

    images = generate(batch_size=16)
    images = images.squeeze().detach().cpu().numpy()
    plt.figure(figsize=(6, 6))
    for i, image in enumerate(images):
        plt.subplot(4, 4, i+1)
        plt.imshow(image, cmap='binary_r')
        plt.axis('off')
    plt.tight_layout()
    plt.show()
